scripts/SREP/cluster-operator-status/script.sh

```diff
@@ -0,0 +1,114 @@
+# degraded/unavailable OpenShift cluster operators.
+#
+# The script can specifically target a cluster operator through
+# the use of a "CLUSTER_OPERATOR" parameter
+
+set -e
+set -o nounset
+set -o pipefail
+
+
+# print_header prints a formatted section break header
+print_header() {
+    clusteroperator=$1
+    heading=$2
+    echo -e "--- [ CLUSTER OPERATOR '${clusteroperator}' ${heading} ] ---\n"
+}
+
+# get_summary prints the basic status for the given cluster operator parameter
+get_summary() {
+    clusteroperator=$1
+
+    # shellcheck disable=SC2016
+    available=$(jq -r --arg CLUSTER_OPERATOR "${clusteroperator}" '.items[] | select(.metadata.name == $CLUSTER_OPERATOR) | .status.conditions[] | select(.type == "Available") | .status' <<< "$OPERATOR_REPORT")
+    # shellcheck disable=SC2016
+    progressing=$(jq -r --arg CLUSTER_OPERATOR "${clusteroperator}" '.items[] | select(.metadata.name == $CLUSTER_OPERATOR) | .status.conditions[] | select(.type == "Progressing") | .status' <<< "$OPERATOR_REPORT")
+    # shellcheck disable=SC2016
+    degraded=$(jq -r --arg CLUSTER_OPERATOR "${clusteroperator}" '.items[] | select(.metadata.name == $CLUSTER_OPERATOR) | .status.conditions[] | select(.type == "Degraded") | .status' <<< "$OPERATOR_REPORT")
+
+    print_header "${clusteroperator}" "SUMMARY"
+    echo " - AVAILABLE:   ${available}"
+    echo " - DEGRADED:    ${degraded}"
+    echo " - PROGRESSING: ${progressing}"
+    echo -e "\n"
+}
+
+# get_description prints a detailed status description for the given cluster 
+# operator parameter
+get_description() {
+    clusteroperator=$1
+
+    if ! co_desc_report=$(oc describe clusteroperator "${clusteroperator}"); then
+       echo "An error was returned when describing ClusterOperator '${clusteroperator}': ${co_desc_report}"
+       return
+    fi
+    print_header "${clusteroperator}" "DESCRIPTION"
+    echo -e "${co_desc_report}\n\n"
+}
+
+# get_logs prints the last 15 lines of logs from the cluster operator pod
+get_logs() {
+    clusteroperator=$1
+
+    # Look up all namespaces referenced in status
+    # shellcheck disable=SC2016
+    relatedns=$(jq -r --arg CLUSTER_OPERATOR "${clusteroperator}" '.items[] | select(.metadata.name == $CLUSTER_OPERATOR) | .status.relatedObjects[] | select(.resource == "namespaces") | .name' <<< "$OPERATOR_REPORT")
+    while IFS= read -r ns; do
+        ns_report=$(oc get deployment --ignore-not-found -n "$ns" -o name)
+        while IFS= read -r deployment; do
+            # Look for any deployment which matches an operator deployment.
+            # The cluster operator name may or may not have the word operator in it.
+            # The cluster operator namespaces may or may not unrelated operator pods in them.
+            if [[ "${deployment}" =~ ${clusteroperator} && "${deployment}${clusteroperator}" =~ operator ]]; then
+                deploymentlogs=$(oc logs "${deployment}" -n "${ns}" --tail=15)
```

Comment:
1. For now, the default is last 15 lines. It may or may not be enough to convey the issue. Reason being, the CO degraded alert may fire after 10-20 minutes and the logs might differ for operators. What about adding an environment variable such as `LOG_FLAGS` which the end user can customise and grep as required if the last 15 lines aren't helpful? Example:
```bash
deploymentlogs=$(oc logs "${deployment}" -n "${ns}" "${LOG_FLAGS}")
```

2. Something which can be extended further is to parse for certain common strings for any issue. Example, strings like "error", "fatal", "timeout", "fail" etc. convey a possible hint from the logs. In this way, we could have `--since=5m` logs and `grep -iE "${LOG_PATTERN}"` where `LOG_PATTERN` could be some strings the user expects. Though, this is heavily dependent on what a particular CO logs but has chances of increasing relevance in output by script. Thoughts?

Reply:
but that the flag will have a default of `--tail=15`

Reply:
Yeah, this is a good suggestion, thanks. I've implemented this now. 
`LOG_PATTERN` will be appended onto the `oc logs` call. This lets you do things such as

`LOG_PATTERN="--since=5m"`
or throw a pipe on the end, ie.
`LOG_PATTERN="| grep -i error"`

If not specified at all, it will default to `--tail 15`.

Reply:
On reflection, I'm a little bit nervous about my current implementation which essentially allows anyone to potentially supply an arbitrary executable command via `LOG_PATTERN`, that's ripe for potential abuse. I'll rework that tomorrow into two separate parameters, an `OC_LOGS_ARGS` and a `GREP_PATTERN` which is explicitly passed to `grep`.

Reply:
Made another attempt at this.

- `OC_LOGS_SINCE` can be used to define the `since` interval (eg. `30m`). If not specified, last 15 lines of logs are returned. I experimented with allowing generic args (eg. `--since=30m`) be supplied as parameters but backplane seemed to have a problem with argument values that are, in turn, `name=value` pairs - broke its parsing (something I'll follow up)
- `LOG_PATTERN` defines a regular expression that will be used in a grep of the logs returned. If not specified, no grepping is performed. The grepping is done in a safe manner that shouldn't allow sneaky "I'll just slide this command into your pipeline" sort of tricks.